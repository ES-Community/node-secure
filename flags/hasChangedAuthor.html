<p>
    The package original author/owner has been updated. This may indicate
    ownership transfer !
</p>
<blockquote>
    <p>
        ⚠️ this flag is not yet as revelant as we want because sometimes we fail
        to retrieve the real package owner.
    </p>
</blockquote>
