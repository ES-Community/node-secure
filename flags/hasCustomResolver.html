<p>
    The package has custom dependencies resolver such as <code>+git</code> or
    <code>+ssh</code> or a local file with <code>file:</code>. Note that pacote
    doesn't support <code>ssh</code> so there is no support in nsecure for this
    kind of resolver.
</p>
<p>
    Documentation:
    <a href="https://docs.npmjs.com/cli/install" rel="nofollow">npm-install</a>
</p>
