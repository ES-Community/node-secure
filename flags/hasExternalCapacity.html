<p>
    The package use a Node.js core package that allow to access the network.
    These core package are:
</p>
<ul>
    <li>http</li>
    <li>https</li>
    <li>net</li>
    <li>http2</li>
    <li>dgram</li>
</ul>
<p>
    This flag only work if the AST analysis as successfully retrieved all
    dependencies as expected.
</p>
