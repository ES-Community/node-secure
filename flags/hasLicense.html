<br />
<p>
    This flag mean that we have not detected any licenses in the npm Tarball (or
    something went wrong in the detection) For detecting licenses we are reading
    the <strong>package.json</strong> and searching for local files that contain
    the word "license".
</p>
<p>
    For more information on how license must be described in the package.json,
    please check the
    <a href="https://docs.npmjs.com/files/package.json#license" rel="nofollow"
        >npm documentation</a
    >.
</p>
<blockquote>
    <p>
        ⚠️ we are working to stabilize this flag !
    </p>
</blockquote>
