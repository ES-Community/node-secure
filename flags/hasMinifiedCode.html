<p>
    Has one or many files that has been detected as minified JavaScript code. We
    use a package that will tell us if the code is minified (in case the file as
    a <strong>.min</strong> then we will consider the file minified by default).
</p>
<br />
<p>
    Minified JavaScript code are commonly used by hacker to obfuscate the code
    to avoid being spotted. A good practice is surely to check all the packages
    with the flag.
</p>
<br />
<p>Example of minified code:</p>
<br />
<pre><code>function cleanRange(version){const firstChar=version.charAt(0);if(firstChar==="^"||firstChar==="<"||firstChar===">"||firstChar==="="||firstChar==="~"){return version.slice(version.charAt(1)==="="?2:1)}
   return version}</code></pre>
<blockquote>
    <p>
        ⚠️ sometimes one line file are considered minified (we are working to
        fix this in the future).
    </p>
</blockquote>
