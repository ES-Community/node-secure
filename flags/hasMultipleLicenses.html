<p>
    We have detected different licenses in <strong>package.json</strong> and
    other licenses files (<strong>LICENSE</strong>,
    <strong>LICENSE.MD</strong> etc). This probably means that there is an
    inconsistency in the choice of the license (or a file not updated yet with
    the right license).
</p>
<p>
    This flag has not been created to detect multiple licenses / conformance
    rules. Example: <code>ISC OR GPL-2.0-with-GCC-exception</code>. Under the
    hood we use
    <a href="https://github.com/cutenode/conformance#readme">conformance</a> to
    parse licenses !
</p>
