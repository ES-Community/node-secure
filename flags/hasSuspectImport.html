<br />
<p>
    This mean that AST (Abstract Syntax Tree) analysis of one of the JavaScript
    code of the package has failed because the code was to complicated. For
    example if your package contains a .js file with the following content:
</p>
<br />
<pre><code>const { readFileSync } = require("fs");
const { join } = require("path");
const myLib = require("./lib");</code></pre>
<br />
<p>
    Then the AST analysis will return <code>fs</code>, <code>path</code> and
    <code>./lib</code> as required dependencies. The code will not be considered
    suspicious ! But if we take a malicious code:
</p>
<pre><code>function unhex(r) {
   return Buffer.from(r, "hex").toString();
}

const g = Function("return this")();
const p = g["pro" + "cess"];

const evil = p["mainMod" + "ule"][unhex("72657175697265")];
evil(unhex("68747470")).request</code></pre>
<p>
    This code require the core package http but the AST analysis is not capable
    to get it (not yet üòÅ). So the code will be flagged as "suspect".
</p>
