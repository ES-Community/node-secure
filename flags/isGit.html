<p>
    The project has been detected as a GIT repository. Sometimes a dependency on
    the package.json link to a GIT repository, example:
</p>
<br/>   
<pre><code>{
    "dependencies": {
        "uNodeHttpServer": "git+https://github.com/tpoisseau/uNodeHttpServer.git#1.2.0"
    }
}</code></pre>
<br/>
<p>
    Because under the hood we use
    <a href="https://github.com/npm/pacote#readme">pacote</a> to fetch and
    extract packages we are supporting this given pattern.
</p>
