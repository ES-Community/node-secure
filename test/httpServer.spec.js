// Import Node.js Dependencies
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import path from "path";

// Import Third-party Dependencies
import test from "tape";
import { get } from "@myunisoft/httpie";
import zup from "zup";
import * as i18n from "@nodesecure/i18n";
import * as flags from "@nodesecure/flags";

// Require Internal Dependencies
import { buildServer } from "../src/http-server/index.js";

// CONSTANTS
const HTTP_PORT = 1337;
const HTTP_URL = new URL(`http://localhost:${HTTP_PORT}`);
const __dirname = path.dirname(fileURLToPath(import.meta.url));

const JSON_PATH = path.join(__dirname, "fixtures", "httpServer.json");
const INDEX_HTML = readFileSync(path.join(__dirname, "..", "views", "index.html"), "utf-8");

// VARS
const httpServer = buildServer(JSON_PATH, {
  port: HTTP_PORT,
  openLink: false
});

test("'/' should return index.html content", async(tape) => {
  const result = await get(HTTP_URL);

  tape.equal(result.statusCode, 200);
  tape.equal(result.headers["content-type"], "text/html");

  const templateStr = zup(INDEX_HTML)({
    lang: i18n.getToken("lang"),
    token: (tokenName) => i18n.getToken(`ui.${tokenName}`)
  });
  tape.equal(result.data, templateStr);

  tape.end();
});

test("'/flags' should return the flags list as JSON", async(tape) => {
  const result = await get(new URL("/flags", HTTP_URL));

  tape.equal(result.statusCode, 200);
  tape.equal(result.headers["content-type"], "application/json;charset=utf-8");

  tape.end();
});

test("'/flags/description/isGit' should return the isGit HTML description", async(tape) => {
  const result = await get(new URL("/flags/description/isGit", HTTP_URL));

  tape.equal(result.statusCode, 200);
  tape.equal(result.headers["content-type"], "text/html");

  tape.equal(result.data, await flags.eagerFetchFlagFile("isGit"));

  tape.end();
});

test("'/flags/description/foobar' should return a 404 error", async(tape) => {
  tape.plan(2);

  try {
    await get(new URL("/flags/description/foobar", HTTP_URL));
  }
  catch (error) {
    tape.equal(error.statusCode, 404);
    tape.equal(error.data.error, "Not Found");
  }

  tape.end();
});

test("'/data' should return the fixture payload we expect", async(tape) => {
  const result = await get(new URL("/data", HTTP_URL));

  tape.equal(result.statusCode, 200);
  tape.equal(result.headers["content-type"], "application/json");
  // tape.equal(result.data).toMatchSnapshot();

  tape.end();
});

test("'/bundle/:name/:version' should return the bundle size", async(tape) => {
  const result = await get(new URL("/bundle/flatstr/1.0.12", HTTP_URL));

  tape.equal(result.statusCode, 200);
  tape.equal(result.headers["content-type"], "application/json;charset=utf-8");
  checkBundleResponse(tape, result.data);
});

test("'/bundle/:name/:version' should return an error if it fails", async(tape) => {
  tape.plan(2);
  const wrongVersion = undefined;
  const wrongPackageName = "br-br-br-brah";

  try {
    await get(new URL(`/bundle/${wrongPackageName}/${wrongVersion}`, HTTP_URL));
  }
  catch (error) {
    tape.equal(error.statusCode, 404);
    tape.equal(error.data.error, "Not Found");
  }

  tape.end();
});

test("'/bundle/:name' should return the bundle size of the last version", async(tape) => {
  const result = await get(new URL("/bundle/flatstr", HTTP_URL));

  tape.equal(result.statusCode, 200);
  tape.equal(result.headers["content-type"], "application/json;charset=utf-8");
  checkBundleResponse(tape, result.data);
});

test("teardown", (tape) => {
  httpServer.server.close();
  tape.end();
});

/**
 * HELPERS
 */

function checkBundleResponse(tapeInstance, payload) {
  tapeInstance.ok(payload.gzip);
  tapeInstance.ok(payload.size);
  tapeInstance.ok(payload.dependencySizes);
}
