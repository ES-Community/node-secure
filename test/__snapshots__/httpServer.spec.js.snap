// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'/data' should return the fixture payload we expect 1`] = `
Object {
  "foo": "bar",
}
`;

exports[`'/flags' should return the flags list as JSON 1`] = `
Object {
  "bannedFile": Object {
    "emoji": "⚔️",
    "title": "hasBannedFile",
    "tooltipDescription": "The project has at least one sensitive file.",
  },
  "customResolver": Object {
    "emoji": "💎",
    "title": "hasCustomResolver",
    "tooltipDescription": "The package has dependencies that are not packages.",
  },
  "duplicated": Object {
    "emoji": "🎭",
    "title": "isDuplicated",
    "tooltipDescription": "The package is also used somewhere else in the dependency tree but with a different version",
  },
  "externalCapacity": Object {
    "emoji": "🌍",
    "title": "hasExternalCapacity",
    "tooltipDescription": "The package use at least one Node.js dependency capable to communicate outside or to establish a listening server.",
  },
  "hasScript": Object {
    "emoji": "📦",
    "title": "hasScript",
    "tooltipDescription": "has \`post\` and/or \`pre\` (un)install npm script",
  },
  "indirectDependencies": Object {
    "emoji": "🌲",
    "title": "hasIndirectDependencies",
    "tooltipDescription": "The package have indirect dependencies.",
  },
  "isDead": Object {
    "emoji": "💀",
    "title": "isDead",
    "tooltipDescription": "The dependency has not received update from at least one year!",
  },
  "isDeprecated": Object {
    "emoji": "⛔️",
    "title": "isDeprecated",
    "tooltipDescription": "The package is deprecated.",
  },
  "isGit": Object {
    "emoji": "☁️",
    "title": "isGit",
    "tooltipDescription": "The package (project) is a git repository",
  },
  "manyPublishers": Object {
    "emoji": "👥",
    "title": "hasManyPublishers",
    "tooltipDescription": "The package has several publishers.",
  },
  "minifiedCode": Object {
    "emoji": "🔬",
    "title": "hasMinifiedCode",
    "tooltipDescription": "The package seems to have files that are minified/uglified.",
  },
  "missingOrUnusedDependency": Object {
    "emoji": "👀",
    "title": "hasMissingOrUnusedDependency",
    "tooltipDescription": "A dependency is missing in package.json or a dependency is installed but never used!",
  },
  "multipleLicense": Object {
    "emoji": "📚",
    "title": "hasMultipleLicenses",
    "tooltipDescription": "The package has multiple licenses in multiple files",
  },
  "nativeCode": Object {
    "emoji": "🐲",
    "title": "hasNativeCode",
    "tooltipDescription": "The package use native components (package, file, configuration).",
  },
  "noLicense": Object {
    "emoji": "📜",
    "title": "hasNoLicense",
    "tooltipDescription": "The package does not seem to have a license.",
  },
  "outdated": Object {
    "emoji": "⌚️",
    "title": "isOutdated",
    "tooltipDescription": "The current package version is not equal to the package latest version!",
  },
  "vulnerabilities": Object {
    "emoji": "🚨",
    "title": "vulnerabilities",
    "tooltipDescription": "The package have one or many vulnerabilities.",
  },
  "warnings": Object {
    "emoji": "⚠",
    "title": "hasWarnings",
    "tooltipDescription": "The AST analysis has detected warnings (suspect import, unsafe regex ..)",
  },
}
`;
